// n8n-nodes-comfyui-all/dist/nodes/logger.js
"use strict";
/**
 * Simple logger for ComfyUI nodes
 * In n8n context, console.log/warn/error are captured and displayed in the UI
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = exports.LogLevel = void 0;
exports.createLogger = createLogger;
var LogLevel;
(function (LogLevel) {
    LogLevel["DEBUG"] = "debug";
    LogLevel["INFO"] = "info";
    LogLevel["WARN"] = "warn";
    LogLevel["ERROR"] = "error";
})(LogLevel || (exports.LogLevel = LogLevel = {}));
class Logger {
    constructor(context, enabled = true) {
        this.context = context;
        this.enabled = enabled;
    }
    formatMessage(level, message) {
        const timestamp = new Date().toISOString();
        return `[${timestamp}] [${level.toUpperCase()}] [${this.context}] ${message}`;
    }
    debug(message, ...args) {
        if (this.enabled) {
            console.debug(this.formatMessage(LogLevel.DEBUG, message), ...args);
        }
    }
    info(message, ...args) {
        if (this.enabled) {
            console.info(this.formatMessage(LogLevel.INFO, message), ...args);
        }
    }
    warn(message, ...args) {
        if (this.enabled) {
            console.warn(this.formatMessage(LogLevel.WARN, message), ...args);
        }
    }
    error(message, error, ...args) {
        if (this.enabled) {
            console.error(this.formatMessage(LogLevel.ERROR, message), error, ...args);
        }
    }
    /**
     * Create a child logger with additional context
     */
    child(additionalContext) {
        return new Logger(`${this.context}:${additionalContext}`, this.enabled);
    }
}
exports.Logger = Logger;
/**
 * Create a logger instance
 */
function createLogger(context, enabled = true) {
    return new Logger(context, enabled);
}
